pilot/pkg/serviceregistry/federation/controller.go

```diff
@@ -84,34 +98,79 @@ type Options struct {
+	return fmt.Sprintf("svc.%s-imports.local", mesh.Name)
+}
+
+func localDomainSuffix(domainSuffix string) string {
+	if domainSuffix == "" {
+		return "svc.cluster.local"
+	}
+	return "svc." + domainSuffix
+}
+
 // NewController creates a new Aggregate controller
-func NewController(opt Options) *Controller {
+func NewController(opt Options, mesh *v1alpha1.MeshFederation, defaultImportConfig, importConfig *v1alpha1.ServiceImports) *Controller {
 	backoffPolicy := backoff.NewExponentialBackOff()
 	backoffPolicy.MaxElapsedTime = 0
+	localDomainSuffix := localDomainSuffix(opt.DomainSuffix)
+	defaultDomainSuffix := defaultDomainSuffixForMesh(mesh)
+	defaultNameMapper := common.NewServiceImporter(defaultImportConfig, nil, defaultDomainSuffix, localDomainSuffix)
 	return &Controller{
-		discoveryURL:   fmt.Sprintf("%s://%s:%d", common.DiscoveryScheme, opt.EgressService, common.DefaultDiscoveryPort),
-		egressService:  opt.EgressService,
-		egressName:     opt.EgressName,
-		networkAddress: opt.NetworkAddress,
-		useDirectCalls: opt.UseDirectCalls,
-		networkName:    opt.NetworkName,
-		namespace:      opt.Namespace,
-		clusterID:      opt.ClusterID,
-		imports:        map[federationmodel.ServiceKey]*existingImport{},
-		instanceStore:  map[host.Name][]*model.ServiceInstance{},
-		configStore:    opt.ConfigStore,
-		xdsUpdater:     opt.XDSUpdater,
-		resyncPeriod:   opt.ResyncPeriod,
-		backoffPolicy:  backoffPolicy,
-	}
+		discoveryURL:        fmt.Sprintf("%s://%s:%d", common.DiscoveryScheme, opt.EgressService, common.DefaultDiscoveryPort),
+		egressService:       opt.EgressService,
+		egressName:          opt.EgressName,
+		networkAddress:      opt.NetworkAddress,
+		useDirectCalls:      opt.UseDirectCalls,
+		namespace:           opt.Namespace,
+		localClusterID:      opt.LocalClusterID,
+		localNetwork:        opt.LocalNetwork,
+		clusterID:           opt.ClusterID,
+		network:             opt.Network,
+		localDomainSuffix:   localDomainSuffix,
+		defaultDomainSuffix: defaultDomainSuffix,
+		defaultNameMapper:   defaultNameMapper,
+		importNameMapper:    common.NewServiceImporter(importConfig, defaultNameMapper, defaultDomainSuffix, localDomainSuffix),
+		imports:             map[federationmodel.ServiceKey]*existingImport{},
+		serviceStore:        map[host.Name]*model.Service{},
+		instanceStore:       map[host.Name][]*model.ServiceInstance{},
+		kubeClient:          opt.KubeClient,
+		configStore:         opt.ConfigStore,
+		xdsUpdater:          opt.XDSUpdater,
+		resyncPeriod:        opt.ResyncPeriod,
+		backoffPolicy:       backoffPolicy,
+		logger:              common.Logger.WithLabels("component", "federation-registry"),
+	}
+}
+
+func (c *Controller) UpdateImportConfig(importConfig *v1alpha1.ServiceImports) {
+	func() {
+		c.storeLock.Lock()
+		defer c.storeLock.Unlock()
+		c.importNameMapper = common.NewServiceImporter(importConfig, c.defaultNameMapper, c.defaultDomainSuffix, c.localDomainSuffix)
+	}()
+	c.resync()
```

Comment:
Is there any risk of this be so quick that the annon function above doesn't have time to finish?

Reply:
it's not `go func()...`, so it will be executed like a normal inline function call.  i did it this way to use the `defer` functionality to ensure the lock is freed (basically, nesting scope).